File: services/participant-ws-gateway/internal/topics/topics.go
// Package topics defines the subject conventions for the participant
// WS gateway. Mirrors services/edge-pubsub/src/topics.ts in Go.
package topics

import (
	"fmt"
	"strings"
)

// Topic names for audience-side pub/sub.
const (
	Participant = "participant"
	Poll        = "poll"
	WordCloud   = "word_cloud"
	QA          = "qa"
	Quiz        = "quiz"
	Reaction    = "reaction"
	Nav         = "nav"
	Sentiment   = "sentiment"
	RaiseHand   = "raise_hand"
	Lifecycle   = "lifecycle"
	Whisper     = "whisper"
)

// For builds a subject for a given session + topic + (optional) shard.
// Examples:
//
//	For("sess-1", Poll, 5)         → "realtime.session.sess-1.poll.shard.5"
//	For("sess-1", Lifecycle, -1)   → "realtime.session.sess-1.lifecycle"
func For(sessionID, topic string, shard int) string {
	if shard < 0 {
		return fmt.Sprintf("realtime.session.%s.%s", sessionID, topic)
	}
	return fmt.Sprintf("realtime.session.%s.%s.shard.%d", sessionID, topic, shard)
}

// ShardFrom extracts the shard index from a topic, or -1 if not present.
func ShardFrom(subject string) int {
	const suffix = ".shard."
	idx := strings.LastIndex(subject, suffix)
	if idx < 0 {
		return -1
	}
	n := 0
	for _, c := range subject[idx+len(suffix):] {
		if c < '0' || c > '9' {
			return -1
		}
		n = n*10 + int(c-'0')
		if n > 1<<20 {
			return -1
		}
	}
	return n
}

// SessionFrom extracts the session id from a topic.
func SessionFrom(subject string) string {
	const prefix = "realtime.session."
	if !strings.HasPrefix(subject, prefix) {
		return ""
	}
	rest := subject[len(prefix):]
	dot := strings.Index(rest, ".")
	if dot < 0 {
		return ""
	}
	return rest[:dot]
}

File: services/edge-pubsub/src/topics.ts
/**
 * @domio/edge-pubsub — topic conventions.
 *
 * Phase 16 W1. Mirrors `services/realtime-gateway/internal/topics`
 * but for the audience side. Subjects are hierarchical and bound to
 * the shard so the WS gateway can subscribe to one shard without
 * pulling the others.
 *
 *   realtime.session.{session_id}.participant  — join/leave/rejoin
 *   realtime.session.{session_id}.poll         — poll_vote events
 *   realtime.session.{session_id}.word_cloud   — word-cloud submits
 *   realtime.session.{session_id}.qa           — qa_submit events
 *   realtime.session.{session_id}.quiz         — quiz_answer events
 *   realtime.session.{session_id}.reaction     — reaction emojis
 *   realtime.session.{session_id}.nav          — nav_vote
 *   realtime.session.{session_id}.sentiment    — sentiment_vote
 *   realtime.session.{session_id}.raise_hand   — raise-hand queue
 *   realtime.session.{session_id}.lifecycle    — started/idle/ended
 *   realtime.session.{session_id}.whisper      — presenter-only
 *
 * Sharded variants prepend `.shard.{N}` so the gateway can route.
 */

export type AudienceTopic =
  | 'participant'
  | 'poll'
  | 'word_cloud'
  | 'qa'
  | 'quiz'
  | 'reaction'
  | 'nav'
  | 'sentiment'
  | 'raise_hand'
  | 'lifecycle'
  | 'whisper';

export function topicFor(input: {
  session_id: string;
  topic: AudienceTopic;
  shard_index?: number;
}): string {
  const base = `realtime.session.${input.session_id}.${input.topic}`;
  if (typeof input.shard_index === 'number') {
    return `${base}.shard.${input.shard_index}`;
  }
  return base;
}

export function shardFromTopic(topic: string): number | null {
  const m = /\.shard\.(\d+)$/.exec(topic);
  if (!m || !m[1]) return null;
  return Number(m[1]);
}

export function sessionFromTopic(topic: string): string | null {
  const m = /^realtime\.session\.([^.]+)\./.exec(topic);
  if (!m || !m[1]) return null;
  return m[1];
}
